### ISSUE TITLE
[iOS] Authenticate is not returning to app after authorizing on Spotify App

### ISSUE THREAD

**Issue by RamonAcedoM** [cjam/react-native-spotify-remote#130]
Hi, I've been trying to debug this issue for 3 days now. On iOS when you call authenticate, the Spotify app will open with no problems, a tick mark will appear that authorization worked, after that, Spotify doesn't redirect back to the application.

If you need something else to narrow down the issue please ask! 

Thanks @cjam for this library!

React-Native Version: 0.64
react-native-spotify-remote: 0.3.7

**Comment by cjam** [comment]
Hey @RamonAcedoM, can you confirm that you have configured your info.plist according the the section described on this page:

https://developer.spotify.com/documentation/ios/quick-start/

i.e. make sure that you have registered an auth callback uri in your plist that is the same one that is registered to your app in the developer.spotify.com console.

**Comment by RamonAcedoM** [comment]
Hi @cjam, yeah, It is working fine on Android, I followed Spotify docs and everything seems correctly setup:

- Callback on spotify dashboard
- Swap / Refresh token server
- Info.plist setup

I don't know if it is related to this:

https://github.com/spotify/ios-sdk/issues/176#issuecomment-609079358

**Comment by cjam** [comment]
hmm interesting.  Have you tried placing the URL handling call in the Scene Delegation handler as per the issue you referenced?

**Comment by RamonAcedoM** [comment]
I can try that, but my project is purely Objective-C, pretty much what RN 0.64 gives you out of the box.

**Comment by cjam** [comment]
gotcha, I'll try to take a look at it when I have a chance

**Comment by RamonAcedoM** [comment]
I've tried to implement that but didn't work. Guess that only applies if you set iOS 13 as your deployment target. What worked is inverting the names of the callback, for some reason myappname://spotify-login-callback was being listed by uri-scheme as spotify-login-callback://. 

Regardless of Spotify being able to redirect to the application, the app is still killed when trying to authenticate() it works sometimes other times just crashes. I've tried having Spotify playing something, Foregrounded or in the background and still crashes with NSInvalidArgument. Swap/refresh endpoints return the correct information. 

Any ideas? @cjam Btw. Android works flawlessly.

**Comment by cjam** [comment]
hmm.. can you share your app delegate code in here?  As well as your redacted apiConfig?

**Comment by RamonAcedoM** [comment]
Sure!

AppDelegate.m
`#import "AppDelegate.h"

#import <React/RCTBridge.h>
#import <React/RCTBundleURLProvider.h>
#import <React/RCTRootView.h>
#import "RNSplashScreen.h"
#import <RNSpotifyRemote.h>
#import <React/RCTLinkingManager.h>
#import <FBSDKCoreKit/FBSDKCoreKit.h>

#ifdef FB_SONARKIT_ENABLED
#import <FlipperKit/FlipperClient.h>
#import <FlipperKitLayoutPlugin/FlipperKitLayoutPlugin.h>
#import <FlipperKitUserDefaultsPlugin/FKUserDefaultsPlugin.h>
#import <FlipperKitNetworkPlugin/FlipperKitNetworkPlugin.h>
#import <SKIOSNetworkPlugin/SKIOSNetworkAdapter.h>
#import <FlipperKitReactPlugin/FlipperKitReactPlugin.h>

static void InitializeFlipper(UIApplication *application) {
  FlipperClient *client = [FlipperClient sharedClient];
  SKDescriptorMapper *layoutDescriptorMapper = [[SKDescriptorMapper alloc] initWithDefaults];
  [client addPlugin:[[FlipperKitLayoutPlugin alloc] initWithRootNode:application withDescriptorMapper:layoutDescriptorMapper]];
  [client addPlugin:[[FKUserDefaultsPlugin alloc] initWithSuiteName:nil]];
  [client addPlugin:[FlipperKitReactPlugin new]];
  [client addPlugin:[[FlipperKitNetworkPlugin alloc] initWithNetworkAdapter:[SKIOSNetworkAdapter new]]];
  [client start];
}
#endif

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
#ifdef FB_SONARKIT_ENABLED
  InitializeFlipper(application);
#endif
  [FBSDKApplicationDelegate initializeSDK:launchOptions]; // <- add this
  RCTBridge *bridge = [[RCTBridge alloc] initWithDelegate:self launchOptions:launchOptions];
  RCTRootView *rootView = [[RCTRootView alloc] initWithBridge:bridge
                                                   moduleName:@"PopExtremo"
                                            initialProperties:nil];

  if (@available(iOS 13.0, *)) {
      rootView.backgroundColor = [UIColor systemBackgroundColor];
  } else {
      rootView.backgroundColor = [UIColor whiteColor];
  }

  self.window = [[UIWindow alloc] initWithFrame:[UIScreen mainScreen].bounds];
  UIViewController *rootViewController = [UIViewController new];
  rootViewController.view = rootView;
  self.window.rootViewController = rootViewController;
  [self.window makeKeyAndVisible];
  [[FBSDKApplicationDelegate sharedInstance] application:application
                             didFinishLaunchingWithOptions:launchOptions];
  [RNSplashScreen show];
  return YES;
}

- (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
{
#if DEBUG
  return [[RCTBundleURLProvider sharedSettings] jsBundleURLForBundleRoot:@"index" fallbackResource:nil];
#else
  return [[NSBundle mainBundle] URLForResource:@"main" withExtension:@"jsbundle"];
#endif
}


- (BOOL)application:(UIApplication *)app openURL:(NSURL *)url options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options {
    if ([[FBSDKApplicationDelegate sharedInstance] application:app openURL:url options:options]) {
      return YES;
    }
    if ([[RNSpotifyRemoteAuth sharedInstance] application:app openURL:url options:options]) {
       return YES;
    }
    if ([RCTLinkingManager application:app openURL:url options:options]){
       return YES;
    }
    return YES;
}

@end
`

ApiConfig:
`        const config: ApiConfig = {
            clientID: SPOTIFY_CLIENT_ID,
            redirectURL: SPOTIFY_REDIRECT_URL,
            tokenRefreshURL: SPOTIFY_TOKEN_REFRESH_URL,
            tokenSwapURL: SPOTIFY_TOKEN_SWAP_URL,
            scopes: [ApiScope.AppRemoteControlScope],
            playURI,
            showDialog,
            authType,
};`

Vars:
`SPOTIFY_CLIENT_ID=d14...
SPOTIFY_REDIRECT_URL=myapp://spotify-login-callback
SPOTIFY_TOKEN_REFRESH_URL=https://mydomain.com/user/refresh
SPOTIFY_TOKEN_SWAP_URL=https://mydomain.com/swap
SPOTIFY_CLIENT_SECRET=3f...`

**Comment by cjam** [comment]
Hmm.. Everything looks good there.  When you're running the app from XCode, do you see any errors or log messages when you try to authenticate that might help narrow down what's going on?

**Comment by RamonAcedoM** [comment]
The issue I intermittently get is a null session object, so it crashes trying to assing null to the session.

**Comment by RamonAcedoM** [comment]
@cjam Well, after the update to 0.3.9 is not crashing anymore, but now I'm not getting the refresh token in the /refresh_token endpoint.

**Comment by krishan-kumar-mourya** [comment]
@RamonAcedoM did you get any solution to return back to the app after authentication?

**Comment by cjam** [comment]
@RamonAcedoM did you ever find a work around / solution to this problem?

**Comment by lucianobracco-geojam** [comment]
@krishan-kumar-mourya  @RamonAcedoM  did you find a solution to this?

**Comment by manisha-am** [comment]
@RamonAcedoM did you find any solution for this happening with me when trying [spotify sdk for flutter](https://pub.dev/packages/spotify_sdk)?

**Comment by krishan-kumar-mourya** [comment]
> @krishan-kumar-mourya @RamonAcedoM did you find a solution to this?

Yes, I was able to fix it. Don't remember the exact fix but I made some changes for Schemes in Info.plist.
My Info.plist has these lines related to the fix.

```
<key>CFBundleSignature</key>
<string>????</string>
<key>CFBundleURLTypes</key>
<array>
<dict>
	<key>CFBundleURLName</key>
	<string>$(PRODUCT_BUNDLE_IDENTIFIER)</string>
	<key>CFBundleURLSchemes</key>
	<array>
		<string>My Demo App</string>
	</array>
</dict>
</array>
<key>CFBundleVersion</key>
<string>1</string>
<key>LSApplicationQueriesSchemes</key>
<array>
<string>spotify</string>
</array>
<key>LSRequiresIPhoneOS</key>
<true/>
```
